### PR TITLE
Spend Permissions docs

### PR DESCRIPTION
**Pull request by amiecorso**
First pass on docs for SpendPermissions.  Overview page, quickstart comprehensive tutorial building a demo app.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: docs/pages/guides/spend-permissions/overview.mdx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    uint160 allowance;
+    uint48 period;
+    uint48 start;
+    uint48 end;
+    uint256 salt;
+    bytes extraData;
+}
+```
+
+
```
Comment (eric-ships) [inline docs/pages/guides/spend-permissions/overview.mdx:35]: Consider adding a section for requesting spend permissions? Maybe the details above could into it.

File: docs/pages/guides/spend-permissions/quick-start.mdx [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+to connect their smart wallet to the application.
+
+From here, we'll modify the app to assemble, sign, approve, and use a spend permission to spend our users' funds.
+
+### Set up some environment variables
+
+<Callout type="danger">
+Always secure your private keys appropriately!  
+We use an environment variable in this demo only for simplicity.
+</Callout>
```
Comment (eric-ships) [inline docs/pages/guides/spend-permissions/quick-start.mdx:36]: 👌🏼

File: docs/pages/guides/spend-permissions/quick-start.mdx [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+For this toy example, go ahead and create a local `.env` file, and add two environment variables: `SUBSCRIPTION_PRIVATE_KEY` and `NEXT_PUBLIC_SUBSCRIPTION_SPENDER`.
```
Comment (eric-ships) [inline docs/pages/guides/spend-permissions/quick-start.mdx:38]: "toy"? I haven't heard of that word used in this context before

**Review by eric-ships — COMMENTED** [review]
Looks great @amiecorso - will let others approve 👍🏼

File: docs/pages/guides/spend-permissions/overview.mdx [hunk 4]
```diff
@@ -0,0 +1,70 @@
+# Spend Permissions
+
+Spend Permissions enable third-party signers to spend assets (native and ERC-20 tokens) from user wallets.
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/overview.mdx:3]: nit: "users' wallets"

File: docs/pages/guides/spend-permissions/overview.mdx [hunk 5]
```diff
@@ -0,0 +1,70 @@
+# Spend Permissions
+
+Spend Permissions enable third-party signers to spend assets (native and ERC-20 tokens) from user wallets. 
+This unlocks the following use cases: // TODO make note about reduced signing
+- frequent in-game spending
+- subscription payments
+- auto-renewals
+- automated or event-based trading
+- micropayment streaming
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/overview.mdx:9]: nit: yeah i'd maybe just say something like

> Once granted, spend permissions allow you to move your users' assets without any kind of signature, unlocking things like subscriptions & trading bots. The reduced signing friction can also be leveraged to enhance UX for high-frequency use cases such as onchain games.

File: docs/pages/guides/spend-permissions/overview.mdx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+`SpendPermissionManager.approveWithSignature` to approve the spend permission onchain.
+
+Multiple spend permissions can be batched and approved with a single signature by constructing a batch spend permission object
+and approving with `SpendPermissionManager.approveBatchWithSignature`.
+
+
+## Revoking
+
+Users can revoke permissions at any time by calling `SpendPermissionManager.revoke`. Calls to revoke can be batched via 
+`CoinbaseSmartWallet.executeBatch`. Once a spend permission has been revoked, it can never be re-approved.
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/overview.mdx:49]: the wallet will call bundle app-defined calls into a call to `executeBatch` on the account. we should just say something like "revocations are onchain calls that can be batched similar to other 4337 transactions"

File: docs/pages/guides/spend-permissions/overview.mdx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+
+## Revoking
+
+Users can revoke permissions at any time by calling `SpendPermissionManager.revoke`. Calls to revoke can be batched via 
+`CoinbaseSmartWallet.executeBatch`. Once a spend permission has been revoked, it can never be re-approved. 
+
+If the user wants to re-authorize the spend permission, the spender will need to generate a new spend permission 
+that has a unique hash from the original spend permission. If the details of the new spend permission are identical
+to the revoked permission, the `salt` field of the permission should be used to generate a unique hash.
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/overview.mdx:53]: imo this is a little too technical for the overview section, would make sense as a notice on the getting started page though

File: docs/pages/guides/spend-permissions/quick-start.mdx [hunk 8]
```diff
@@ -0,0 +1,702 @@
...
+
+This will generate an app that is ready to run and contains a wallet connection button that users can use
+to connect their smart wallet to the application.
+
+From here, we'll modify the app to assemble, sign, approve and use a spend permission to spend our users' funds!
+
+
+### Install additional dependencies
+
+We'll need a few additional dependencies: [Viem](https://viem.sh/), [Wagmi](https://wagmi.sh/), and [TanStack](https://tanstack.com/query/latest)
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/quick-start.mdx:27]: onchainkit should probably just ship with this stuff? @fakepixels @0xAlec

File: docs/pages/guides/spend-permissions/quick-start.mdx [hunk 9]
```diff
@@ -0,0 +1,702 @@
...
+Create a local `.env` file and add two environment variables:
+
+```
+SUBSCRIPTION_PRIVATE_KEY=
+NEXT_PUBLIC_SUBSCRIPTION_SPENDER=
+```
+
+These will represent our spender private key and the address of our spender smart contract wallet.
+
+You can generate a random private key [here](https://privatekeys.pw/keys/ethereum/random), or maybe you have a development private key
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/quick-start.mdx:51]: maybe just recommend viem or foundry

File: docs/pages/guides/spend-permissions/quick-start.mdx [hunk 10]
```diff
@@ -0,0 +1,702 @@
...
+Run this script to log the counterfactual address of your new smart contract wallet and assign this address to the 
+`NEXT_PUBLIC_SUBSCRIPTION_SPENDER` variable in your .env.
+
+<Callout type="info">
+You may need to install ts-node or other of the following dependencies: `npm install -g ts-node typescript @types/node dotenv`  
+</Callout>
+
+<Callout type="info">
+You may need to temporarily set the value of `"module"` in your `tsconfig.json` to `"commonjs"` to run this script.
+</Callout>
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/quick-start.mdx:116]: probably just add a code snipped for how to run

File: docs/pages/guides/spend-permissions/quick-start.mdx [hunk 11]
```diff
@@ -0,0 +1,702 @@
...
+### Set up your spender wallet
+
+<Callout type="danger">
+Always secure your private keys appropriately!  
+We insecurely use an environment variable in this demo for simplicity.
+</Callout>
+
+Create a local `.env` file and add two environment variables:
+
+```
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/quick-start.mdx:44]: would mention `NEXT_PUBLIC_CDP_API_KEY` here too

Comment (lukasrosario) [reply docs/pages/guides/spend-permissions/quick-start.mdx:44]: `BASE_SEPOLIA_PAYMASTER_URL` too

File: docs/pages/guides/spend-permissions/quick-start.mdx [hunk 12]
```diff
@@ -0,0 +1,702 @@
...
+You'll need one more environment variable, which is `BASE_SEPOLIA_PAYMASTER_URL`.  
+This one's easy if you already have your CDP API key:  
+  `"https://api.developer.coinbase.com/rpc/v1/base-sepolia/{YOUR_CDP_API_KEY}"`
+
+
+### Create a spender client
+
+Our client is what our app will use to communicate with the blockchain. In this example, our client is a Coinbase Smart Wallet, 
+and we'll use a [paymaster](/guides/paymasters.mdx) to sponsor our transactions so we don't have to worry
+about having gas money.
```
Comment (lukasrosario) [inline docs/pages/guides/spend-permissions/quick-start.mdx:133]: nit: somewhat misleading as we will have to pay CDP back 😅 maybe "so we don't have to worry about having ETH in the spender account"

**Review by ilikesymmetry — APPROVED** [review]

File: docs/pages/guides/spend-permissions/quick-start.mdx [hunk 13]
```diff
@@ -0,0 +1,702 @@
...
+
+A complete example of this demo can be found [here](https://github.com/ilikesymmetry/subscribe-onchain).  
+Smart contract deployment addresses for `SpendPermissionManager.sol` can be found [here](https://github.com/coinbase/spend-permissions).
+
+:::steps
+### Set up a basic app template using OnchainKit
+Set up a boilerplate React/Next app by running the following command and following the instructions:
+
+```bash
+npm create onchain
```
[the remark this answers is not in the document]
Comment (0xAlec) [reply docs/pages/guides/spend-permissions/quick-start.mdx:16]: yes, bun works too with our CLI

File: docs/pages/guides/spend-permissions/quick-start.mdx (hunk 9 above)
Comment (0xAlec) [reply docs/pages/guides/spend-permissions/quick-start.mdx:51]: in https://github.com/coinbase/paymaster-bundler-examples we recommend `cast wallet new`

File: docs/pages/guides/spend-permissions/quick-start.mdx (hunk 8 above)
Comment (0xAlec) [reply docs/pages/guides/spend-permissions/quick-start.mdx:27]: these have been added as explicit dependencies here - https://github.com/coinbase/onchainkit/pull/1570

cc: @dschlabach

**Review by ilikesymmetry — APPROVED** [review]

**Pull request merged**